### PR TITLE
Added Support For Apache JSON and Binary Data

### PR DESCRIPTION
**Pull request by JonnoFTW**
#13 Added support for apache json, so that apache thrift server/clients can talk to thriftpy2 server/clients.
#136 Added support for binary data fields, I'm not sure what ramifications this would have on python2.

### CODE REVIEW

**Comment by ethe** [comment]
Some CI has failed, do you mind to fix it?

**Comment by JonnoFTW** [comment]
@ethe don't merge yet, I have some problems in my own codebase where `TApacheJSONProtocol` does not work with `THttpClient`

**Comment by JonnoFTW** [comment]
@ethe feel free to check it out now

File: thriftpy2/__init__.py [hunk 1]
```diff
@@ -5,7 +5,7 @@
 from .hook import install_import_hook, remove_import_hook
 from .parser import load, load_module, load_fp
 
-__version__ = '0.4.11'
+__version__ = '0.4.12'
```
Comment (ethe) [inline thriftpy2/__init__.py:8]: Please do not modify the version, I will upgrade the version when the package releasing.

File: thriftpy2/protocol/__init__.py [hunk 2]
```diff
@@ -23,5 +24,6 @@
 __all__ = ['TBinaryProtocol', 'TBinaryProtocolFactory',
            'TCyBinaryProtocol', 'TCyBinaryProtocolFactory',
            'TJSONProtocol', 'TJSONProtocolFactory',
+           'TApacheJSONProtocol', 'TApacheJSONProtocolFactory',
            'TMultiplexedProtocol', 'TMultiplexedProtocolFactory',
-           'TCompactProtocol', 'TCompactProtocolFactory']
+           'TCompactProtocol', 'TCompactProtocolFactory', ]
```
Comment (ethe) [inline thriftpy2/protocol/__init__.py:29]: Maybe it is better to leave it as before.

File: thriftpy2/thrift.py [hunk 3]
```diff
@@ -104,6 +104,7 @@ class TType(object):
     LIST = 15
     UTF8 = 16
     UTF16 = 17
+    BINARY = 18
```
Comment (ethe) [inline thriftpy2/thrift.py:107]: There was already a BINARY member at https://github.com/Thriftpy/thriftpy2/pull/139/files#diff-b55308698cb9187e83350298d1d941b56d1e956b1e36f36c00fe61e68145cef5R100

File: thriftpy2/protocol/__init__.py (hunk 2 above)
Comment (JonnoFTW) [reply thriftpy2/protocol/__init__.py]: Fixed

File: thriftpy2/thrift.py (hunk 3 above)
Comment (JonnoFTW) [reply thriftpy2/thrift.py:106]: Fixed in 6d1b8c9

File: thriftpy2/__init__.py (hunk 1 above)
Comment (JonnoFTW) [reply thriftpy2/__init__.py]: Fixed in 8bb4ed6

File: thriftpy2/thrift.py (hunk 3 above)
Comment (ethe) [reply thriftpy2/thrift.py:106]: Does it affect other protocols like binary? Maybe we should check it in all protocols.

Comment (ethe) [reply thriftpy2/thrift.py:106]: Other protocols can not treat the new BINARY member if the value of it is not `11` anymore.

**Comment by ethe** [comment]
Good job and thank you, if we can confirm that it does not change features which already exist, I think we can merge it right now.

**Comment by JonnoFTW** [comment]
@ethe I updated asyncio protocols to support binary fields and so far all the tests work. Might be a need to do some interoperability tests with other implementations.

**Review by ethe — APPROVED** [review]

File: setup.py [hunk 4]
```diff
@@ -15,6 +15,7 @@
 
 install_requires = [
     "ply>=3.4,<4.0",
+    "six>=1.15.0"
```
Comment (aisk) [inline setup.py:18]: I thought we should pin the max version of six, for six 2.0 may have breaking changes. `six~=1.15` is suitable here.
and it's better to have comma and the end of line.

**Review by aisk — CHANGES_REQUESTED** [review]

File: thriftpy2/protocol/apache_json.py [hunk 5]
```diff
@@ -0,0 +1,280 @@
...
+        suitable_for_isinstance = (suitable_for_isinstance,)
+    for thing in suitable_for_isinstance:
+        if isinstance(thing, tuple):
+            types.extend(flatten(thing))
+        else:
+            types.append(thing)
+    return tuple(types)
+
+
+class TApacheJSONProtocolFactory:
```
Comment (aisk) [inline thriftpy2/protocol/apache_json.py:56]: All classes should be inherit from object for we have python2 compatible.

File: thriftpy2/protocol/apache_json.py [hunk 6]
```diff
@@ -0,0 +1,280 @@
...
+        return TApacheJSONProtocol(trans)
+
+
+class TApacheJSONProtocol(TProtocolBase):
+    """
+    Protocol that implements the Apache JSON Protocol
+    """
+
+    def __init__(self, trans):
+        TProtocolBase.__init__(self, trans)
```
Comment (aisk) [inline thriftpy2/protocol/apache_json.py:68]: It's better to call `super(TApacheJSONProtocol, self).__init__(trans)`.

File: thriftpy2/protocol/apache_json.py [hunk 7]
```diff
@@ -0,0 +1,280 @@
...
+        pass
+
+    def write_message_begin(self, name, ttype, seqid):
+        self.api = name
+        self.ttype = ttype
+        self.seqid = seqid
+
+    def write_struct(self, obj):
+        """
+        Write json to self.trans following apache style jsonification of `obj`
```
Comment (aisk) [inline thriftpy2/protocol/apache_json.py:117]: Missing new line between comments and type annotations.

File: thriftpy2/protocol/compact.py [hunk 8]
```diff
@@ -108,7 +110,8 @@ class CompactType(object):
     TType.STRUCT: CompactType.STRUCT,
     TType.LIST: CompactType.LIST,
     TType.SET: CompactType.SET,
-    TType.MAP: CompactType.MAP
+    TType.MAP: CompactType.MAP,
+    TType.BINARY: CompactType.BINARY
```
Comment (aisk) [inline thriftpy2/protocol/compact.py:114]: It's better to have a comman here.

File: thriftpy2/protocol/apache_json.py [hunk 9]
```diff
@@ -0,0 +1,280 @@
...
+
+    def read_struct(self, obj):
+        """
+        Read the next struct into obj, usually the argument from an incoming request
+        Only really used to read the arguments off a request into whatever we want
+        see thriftpy2.thrift.TProcessor.process_in for how this class will be used
+
+        Will turn the contents of self.req[4] into the args of obj,
+        ie. self.req[4]["1"] must be rendered into obj.thrift_spec
+        :param obj:
```
Comment (aisk) [inline thriftpy2/protocol/apache_json.py:277]: Same as mentioned above.

File: thriftpy2/protocol/apache_json.py [hunk 10]
```diff
@@ -0,0 +1,280 @@
...
+                else:
+                    result[field_idx] = {
+                        CTYPES[ttype]: val
+                    }
+        return result
+
+    def _dict_to_thrift(self, data, base_type):
+        """
+        Convert an apache thrift dict (where key is the type, value is the data)
+        :param data:
```
Comment (aisk) [inline thriftpy2/protocol/apache_json.py:197]: Same as mentioned above.

File: thriftpy2/protocol/apache_json.py [hunk 11]
```diff
@@ -0,0 +1,280 @@
+# -*- coding: utf-8 -*-
+"""
+Transport for json protocol that apache thrift files will understand
+unfortunately, thriftpy2's TJSONProtocol is not compatible with apache's
+"""
+from __future__ import absolute_import
```
Comment (aisk) [inline thriftpy2/protocol/apache_json.py:6]: Missing newline here.

File: thriftpy2/protocol/apache_json.py [hunk 12]
```diff
@@ -0,0 +1,280 @@
...
+import json
+import base64
+
+from six import string_types
```
Comment (aisk) [inline thriftpy2/protocol/apache_json.py:10]: Missing newline here.

File: tests/test_apache_json.py [hunk 13]
```diff
@@ -0,0 +1,171 @@
+from __future__ import absolute_import
```
Comment (aisk) [inline tests/test_apache_json.py:1]: Missing coding magic line here.

File: thriftpy2/protocol/apache_json.py (hunk 6 above)
Comment (JonnoFTW) [reply thriftpy2/protocol/apache_json.py:68]: I copied the usage from existing protocol classes:

https://github.com/Thriftpy/thriftpy2/blob/17a66b5c3fdef6dd66c5d853c08eafcbffb0b7e4/thriftpy2/protocol/binary.py#L363

https://github.com/Thriftpy/thriftpy2/blob/17a66b5c3fdef6dd66c5d853c08eafcbffb0b7e4/thriftpy2/protocol/compact.py#L127

https://github.com/Thriftpy/thriftpy2/blob/17a66b5c3fdef6dd66c5d853c08eafcbffb0b7e4/thriftpy2/protocol/json.py#L166

Unless you want to change all of them to `super` calls?

File: setup.py (hunk 4 above)
Comment (JonnoFTW) [reply setup.py]: Fixed in d87d5cc

File: tests/test_apache_json.py (hunk 13 above)
Comment (JonnoFTW) [reply tests/test_apache_json.py:1]: Fixed in ffb624c

File: thriftpy2/protocol/apache_json.py (hunk 12 above)
Comment (JonnoFTW) [reply thriftpy2/protocol/apache_json.py:12]: Fixed ffb624c

File: thriftpy2/protocol/apache_json.py (hunk 11 above)
Comment (JonnoFTW) [reply thriftpy2/protocol/apache_json.py:8]: Fixed ffb624c

File: thriftpy2/protocol/apache_json.py (hunk 10 above)
Comment (JonnoFTW) [reply thriftpy2/protocol/apache_json.py:202]: Fixed ffb624c

File: thriftpy2/protocol/apache_json.py (hunk 9 above)
Comment (JonnoFTW) [reply thriftpy2/protocol/apache_json.py:283]: Fixed ffb624c

File: thriftpy2/protocol/apache_json.py (hunk 7 above)
Comment (JonnoFTW) [reply thriftpy2/protocol/apache_json.py:120]: Fixed ffb624c

File: thriftpy2/protocol/apache_json.py (hunk 5 above)
Comment (JonnoFTW) [reply thriftpy2/protocol/apache_json.py]: Fixed ffb624c

File: thriftpy2/protocol/compact.py (hunk 8 above)
Comment (JonnoFTW) [reply thriftpy2/protocol/compact.py]: Fixed 23ab765

File: setup.py [hunk 14]
```diff
@@ -15,7 +15,7 @@
 
 install_requires = [
     "ply>=3.4,<4.0",
-    "six>=1.15.0"
+    "six~=1.15.0",
```
Comment (aisk) [inline setup.py:18]: It should be `six~=1.15`, the `six~=1.15.0` have different meanings. See here: https://www.python.org/dev/peps/pep-0440/#examples

File: thriftpy2/protocol/apache_json.py (hunk 6 above)
Comment (aisk) [reply thriftpy2/protocol/apache_json.py:71]: I think we should not modify the old codes in this PR. Maybe we can change it in the new codes only?

Comment (JonnoFTW) [reply thriftpy2/protocol/apache_json.py:71]: @aisk Feel free to create an issue and I will fix it in a 2nd pull request.

**Comment by aisk** [comment]
> Hi @JonnoFTW I have a question, is the changes for #136 will make a incompatible change for current users for both python2 & 3? If so, I think we should and a flag to enable this feature, for we don't breaking current users' codes in minor changes.

Hi @JonnoFTW what about this question?

File: thriftpy2/thrift.py (hunk 3 above)
Comment (JonnoFTW) [reply thriftpy2/thrift.py:106]: Other protocols have been updated to support the  `BINARY` type

**Comment by JonnoFTW** [comment]
@aisk tests pass for python2 and python3 so I don't think it breaks and existing code since the old tests still pass with the assumption that they have coverage for `STRING` and `BINARY` thrift types.

**Review by ethe — APPROVED** [review]

**Comment by aisk** [comment]
> @aisk tests pass for python2 and python3 so I don't think it breaks and existing code since the old tests still pass with the assumption that they have coverage for `STRING` and `BINARY` thrift types.

Really sorry for the late of reply. I tried to understand the changes for #136 and took a dig and found there may have some breaking changes.

For example, before this PR, if we have a thrift file that returns a binary field, and it's content is valid utf-8 contents like "hello", the user will got an unicode in Python2 (str in Python3), but after this PR, he will got str in Python (bytes in Python3). If the content only contains ASCII, maybe we can use str / unicode mixed up, but for other cases this will brings trouble.

I think maybe our test case does not covered this case, if you have any doubt about this, maybe I can add some tests for this.

But for now, I think it's cool if you can split this PR into two, the changes for #13 looks good to me.

**Comment by JonnoFTW** [comment]
I've created a small test repo to check that my changes don't break backwards compatibility with Apache Thrift's python implementation (and hopefully other versions too).

https://github.com/JonnoFTW/thrift_interop_test/blob/master/tests/test_interop.py

File: thriftpy2/protocol/binary.py [hunk 15]
```diff
@@ -225,6 +228,10 @@ def read_val(inbuf, ttype, spec=None, decode_response=True):
     elif ttype == TType.DOUBLE:
         return unpack_double(inbuf.read(8))
 
+    elif ttype == TType.BINARY:
+        sz = unpack_i32(inbuf.read(4))
+        return inbuf.read(sz)
```
[the remark this answers is not in the document]
Comment (JonnoFTW) [reply thriftpy2/protocol/binary.py:233]: @aisk what do you want me to do then?

I've checked in the apache version and they return `bytes` for a binary type field.

[the remark this answers is not in the document]
Comment (aisk) [reply thriftpy2/protocol/binary.py:233]: I mean thriftpy2 compatibility, previous vedrsion byted field may return unicode.

[the remark this answers is not in the document]
Comment (JonnoFTW) [reply thriftpy2/protocol/binary.py:233]: I ran the following test in python 2.7.18:

```py
from __future__ import print_function
import os
import time
from multiprocessing import Process

import thriftpy2
from thriftpy2.rpc import make_client, make_server

cdir = os.path.dirname(os.path.realpath(__file__))
test_thrift = thriftpy2.load(cdir + "/spec.thrift")
Bar = test_thrift.Bar
Foo = test_thrift.Foo


def test_img():
    with open(cdir + "/test.png", 'rb') as infile:
        data = infile.read()
    obj = Bar(
        tbinary=data,
        tlist_of_binary=[b'a binary string']
    )

    class Handler(object):
        def test(self, t):
            return t

    def do_server():
        server = make_server(service=test_thrift.BarService, handler=Handler(), host='localhost', port=9090)
        server.serve()

    proc = Process(target=do_server)
    proc.start()
    time.sleep(0.1)

    client = make_client(test_thrift.BarService, host='localhost', port=9090)
    res = client.test(obj)
    print("Type of binary=", type(res.tbinary))
    print("Type of t_list_of_binary=", type(res.tlist_of_binary[0]))
    proc.terminate()


if __name__ == "__main__":
    test_img()

```

With `thriftpy2==0.4.10` I get:

```
Type of binary= <type 'str'>
Type of t_list_of_binary= <type 'unicode'>
```

With `thriftpy2==0.4.12` I get:
```
Type of binary= <type 'str'>
Type of t_list_of_binary= <type 'str'>
```

Is this going to break anyone's code?

[the remark this answers is not in the document]
Comment (aisk) [reply thriftpy2/protocol/binary.py:233]: Hi @JonnoFTW , I tried to run your codes on Python2, got the same output with master branch. But the got the output with current PR:

```
Type of binary= <type 'NoneType'>
Type of t_list_of_binary= <type 'NoneType'>
```

But I think maybe you are right, the previous behavior seems more like a bug and nobody should rely on it. What's your opinion @ethe ?

**Review by aisk — APPROVED** [review]

**Comment by aisk** [comment]
Hi @JonnoFTW I approved this PR, thanks!

**Comment by JonnoFTW** [comment]
@ethe are you happy to merge now? All tests pass along with those in my interoperability repo.

**Comment by ethe** [comment]
Merged, thank you for your patience and contribution.

File: thriftpy2/protocol/binary.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (ethe) [reply thriftpy2/protocol/binary.py:233]: @aisk I think it does not break the backward compatibility.

**Comment by dmulter** [comment]
I've been using Thriftpy2 with a private Thrift server for a long time now, and this latest addition has broken things for me. Sorry I can't share the Thrift files, but here's the error when making an RPC call:

```
.virtualenv/lib/python3.8/site-packages/thriftpy2/thrift.py:219: in _req
    return self._recv(_api)
.virtualenv/lib/python3.8/site-packages/thriftpy2/thrift.py:231: in _recv
    fname, mtype, rseqid = self._iprot.read_message_begin()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

>   ???
E   cybin.ProtocolError: No protocol version header

thriftpy2/protocol/cybin/cybin.pyx:454: ProtocolError
```

**Comment by JonnoFTW** [comment]
@dmulter are you able to provide a test case to recreate this? Also did you install from pypi or from git?

**Comment by JonnoFTW** [comment]
Can you try installing from git?

**Comment by dmulter** [comment]
Fails the same way.

**Comment by JonnoFTW** [comment]
@dmulter can you please give a minimum working example or just tell me exactly what combination of client/server protocols and transports you are using on each end. I'll add it to the tests and see if I can recreate the problem.

**Comment by ethe** [comment]
@dmulter could you please provide a sample IDL file? we need to reproduce it.

**Comment by JonnoFTW** [comment]
@dmulter I'm unable to replicate the problem you are having, here's my test case (it still passes if I use `TCompactProtocol`):

Spec:

```
struct BinarySpec {
    1: map<string, binary> str2bin,
    2: map<string, string> str2str,
}

service BinaryService {
    BinarySpec run(1: BinarySpec b_spec);
}
```

```python
def test_tcompact_memory_buffer():
    spec = load("binary_test.thrift")

    factory = cybin.TCyBinaryProtocolFactory()
    obj = spec.BinarySpec(
        str2bin={"key": b"Binary Value"},
        str2str={"key": "String Value"}
    )

    class Handler:
        def run(self, b_spec):
            print("\nServer received:")
            print(b_spec)
            return b_spec

    def run_server():
        server = make_server(
            service=spec.BinaryService,
            handler=Handler(),
            proto_factory=factory,
        )
        server.serve()

    proc = Process(target=run_server)
    try:
        proc.start()
        time.sleep(0.5)
        client = make_client(spec.BinaryService, proto_factory=factory)
        res = client.run(obj)
        assert res
        print("Server sent:")
        print(res)
    finally:
        proc.kill()
    time.sleep(0.2)
```

Gives the output:

```
Server received:
BinarySpec(str2bin={'key': b'Binary Value'}, str2str={'key': 'String Value'})
Server sent:
BinarySpec(str2bin={'key': b'Binary Value'}, str2str={'key': 'String Value'})
```


If you could modify this test so that it fails in the same way your code does, that would be very useful in fixing the issue.

**Comment by dmulter** [comment]
Some additional status:

* I thought I had installed from a Git clone, but just tried as part of my testing and I get the following error during build: `clang: error: no such file or directory: 'thriftpy2/transport/cybase.c'`
* I verified that the test you created above works fine. I had to change things a bit to avoid multiprocessing errors that are also included in the next bullet item code.
* I created a version that makes the specific service call and it also works fine. See code at the end.
* Seems clear to me that the issue requires the actual server implementation since local tests are not exposing the problem. My strategy now is trying to identify why the Thriftpy2-0.4.13 release broke when talking to the server. I'm trying to build a local version that I can use for debugging the returned bytes to help identify what the code is choking on. I first have to solve the first item above. Note that I can't give access to the real server since it is not ours and is both private and protected via a layer of complex authentication that must be performed first. I experimented with various other transports and was not able to find a way to make the test pass.

xxx.thrift:
```python
enum ApiVersion
{
    kApiVersion1 = 1;
    kApiVersion2;
    kApiVersion3;
}
const ApiVersion kServerApiVersion_Current = ApiVersion.kApiVersion2;

service PrivateService
{
    void openSession
        (
        1: string sessionId,
        2: ApiVersion clientVersion
        )
}
```

test_xxx.py:
```python
import time
import sys
from thriftpy2 import load
from thriftpy2.rpc import make_client, make_server
from multiprocessing import Process

class Handler:
    def openSession(self, sessionId, clientVersion):
        print(f"\nServer received: sessionId={sessionId} clientVersion={clientVersion}")
        sys.stdout.flush()

def run_server():
    print("Starting server...")
    server = make_server(
        service=spec.PrivateService,
        handler=Handler(),
    )
    server.serve()

spec = load("xxx.thrift")

def test_private_service():
    proc = Process(target=run_server)
    try:
        proc.start()
        time.sleep(0.5)

        print("Calling openSession...")
        client = make_client(service=spec.PrivateService)
        session_id = "xxx"
        client.openSession(session_id, spec.kServerApiVersion_Current)
        assert False, "tests actually worked" # so we can see print output
    finally:
        proc.kill()
    time.sleep(0.2)
```

**Comment by dmulter** [comment]
Installed Cython and built from source. Currently experimenting with what's returned from the server. Not sure if it helps, but the `size` value it expects for the protocol version header has a value of `3c68746d`. Trying to get the full server response...

**Comment by dmulter** [comment]
Found the problem and it was not this PR, but a different change on 0.4.13. The protocol error was because HTML for an error page was being returned. The breaking change was due to an inserted path separator after port as I'm passing a parsed URL as follows:

```python
    from urllib.parse import urlparse

    url = urlparse(full_url)
    client = make_client(
        service=spec.TestService,
        scheme=url.scheme,
        host=url.netloc,
        path=url.path,
        port="443"
    )
```

And the change was https://github.com/Thriftpy/thriftpy2/pull/148/commits/faf10551188fc23e3293c526e1142abbdf2a9d14

**Comment by JonnoFTW** [comment]
@ethe considering that the @dmulter 's issue didn't come from this pull request, can we merge it back in? The slash problem can be fixed separately.

**Comment by ethe** [comment]
@JonnoFTW yeah your merge does not break compact protocol, however, there is another problem, see #156

**Pull request merged**